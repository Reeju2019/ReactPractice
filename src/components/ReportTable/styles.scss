.mylink {
  text-decoration: underline;
  color: #0d6efd;
}
.ag-theme-alpine {
  .ag-ltr {
    .ag-cell {
      text-align: left;
    }
  }
  .ag-row {
    // border: 0;
    .ag-ltr {
      .ag-cell {
        text-align: left;
      }
    }
  }
}
.ag-theme-alpine {
  .ag-root-wrapper {
    border: 0;
    .ag-header {
      // border: 0;
    }
  }
}

.strong {
  font-weight: bold;
}

.link-row .ag-cell:last-child {
  text-decoration: underline;
  color: #0d6efd;
  cursor: pointer;
}

.no-link-row .ag-cell:last-child {
  text-decoration: none;
  color: black;
}

.ag-header-viewport {
  background-color: white;
}

.rchart-container-border {
  border: #dcdcdc solid;
  width: 50%;
}

.rchart-border-bottom {
  border-bottom: #dcdcdc solid;
  margin: 0;
}
