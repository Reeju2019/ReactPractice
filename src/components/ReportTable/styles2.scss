.mylink {
  text-decoration: underline;
  color: #0d6efd;
}
.ag-theme-alpine {
  .ag-ltr {
    .ag-cell {
      text-align: left;
    }
  }
  .ag-row {
    border: 0;
    .ag-ltr {
      .ag-cell {
        text-align: left;
      }
    }
  }
}
.ag-theme-alpine {
  .ag-root-wrapper {
    border: 0;
    .ag-header {
      border: 0;
    }
  }
}

.strong {
  font-weight: bold;
}

.ag-header-viewport {
  background-color: white;
}

.ag-row .ag-cell:last-child {
  text-decoration: none;
  color: black;
}
