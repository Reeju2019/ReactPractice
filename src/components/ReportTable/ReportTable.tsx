import React from "react";
import { Col, FormLabel, Row, Image, Form } from "react-bootstrap";
import { AgGridReact } from "ag-grid-react";

import "ag-grid-community/dist/styles/ag-grid.css";
import "ag-grid-community/dist/styles/ag-theme-alpine.css";

import "./styles.scss";
import ReactTooltip from "react-tooltip";

/**
 * @description Props of ReportTable Component
 * @memberof ReportTable
 * @property {any} localeConfig Set the locale config data
 * @property {any} reportTableConfig Set the report table config data
 * @property {any} viewDetails Set the view details URL
 * @property {string} header Set the header text
 * @property {string} headerLogo Set the header logo URL
 * @property {any} layout Set the layout
 * @property {any} dropdownHandler Custom function to handler dropdown
 * @property {number} height Set the height of the table
 * @property {booblean} isLink Set true if their is a link in 4th column else set to false
 */
/* eslint-disable */
interface IProps {
  localeConfig: any;
  reportTableConfig: any;
  viewDetails: any;
  header: string;
  headerLogo: string;
  layout: any;
  dropdownHandler: any;
  height?: number;
  isLink?: boolean;
}

/**
 * @description States of ReportTable Component
 * @memberof ReportTable
 * @property {any} reportTableConfig Set the report table config data
 * @property {number} activeRowData Set the active row index
 * @property {number} recentRowSelection Set the recent row selection index
 */
/* eslint-disable */
interface IState {
  reportTableConfig: any;
  activeRowData: number;
  recentRowSelection: number;
}

/**
 * @component
 * This component was used in App component
 * @description ReportTable component
 */
class ReportTable extends React.Component<IProps, IState> {
  HEADER_HEIGHT = 65;
  CELL_HEIGHT = 43;

  state: IState = {
    reportTableConfig: null,
    activeRowData: 0,
    recentRowSelection: -1,
  };

  /**
   * @constructor
   * @description ReportTable component
   * @param {IProps} props Props
   * @returns {void}
   */
  constructor(props: any) {
    super(props);
    this.bindThisToMethods = this.bindThisToMethods.bind(this);
    this.bindThisToMethods();
  }

  /**
   * @description Bind all the methods to this
   * @property {Function} bindThisToMethods Bind This To Methods
   * @returns void
   */
  bindThisToMethods() {
    this.setConfig = this.setConfig.bind(this);
    this.rowClickHandler = this.rowClickHandler.bind(this);
    this.tableHeight = this.tableHeight.bind(this);
  }

  /**
   * @description Veturns the height of the table
   * @property {function} tableHeight Table Height
   * @memberof ReportTable
   * @returns {number}
   */
  tableHeight() {
    const rowsCount = this.props.reportTableConfig.rowData.length;
    return this.HEADER_HEIGHT + rowsCount * this.CELL_HEIGHT;
  }

  /**
   * @description Set state according to the json configuration
   * @property {Function} setConfig Set Config
   * @returns {void}
   */
  setConfig() {
    const data = this.props.reportTableConfig;
    const newData = {
      ...data,
      columnDefs: data.columnDefs.map((col: any) => {
        return {
          ...col,
          headerTooltip: col.headerTooltip ?? col.headerName,
        };
      }),
    };
    this.setState({ reportTableConfig: newData });
  }

  /**
   * @description Handles the row click event
   * @property {function} rowClickHandler Row Click Handler
   * @param {Event} CellClickedEvent Cell Clicked Event
   * @memberof ReportTable
   * @returns {void}
   */
  rowClickHandler(CellClickedEvent) {
    const rowIndex = CellClickedEvent.node.rowIndex;

    this.state.recentRowSelection === rowIndex
      ? CellClickedEvent.event.target.parentElement.classList.remove(
          "ag-row-selected"
        )
      : CellClickedEvent.event.target.parentElement.classList.add(
          "ag-row-selected"
        );

    this.setState({
      recentRowSelection:
        this.state.recentRowSelection === rowIndex ? -1 : rowIndex,
    });

    this.props.viewDetails(CellClickedEvent, this.props.isLink ?? true);
  }

  /**
   * @description The state will be set to the json configuration here
   * @property {function} componentDidMount Component Did Mount
   * @memberof ReportTable
   */
  componentDidMount() {
    this.setConfig();
  }

  render() {
    const getRowClass = (params: any) => {
      return this.props.isLink ?? true ? "link-row" : "no-link-row";
    };
    return (
      <>
        {this.state.reportTableConfig && (
          <Row className="rchart-container-border m-5 m-5">
            <Row className="rchart-border-bottom py-2 ps-4">
              <Col
                xs={this.props.layout.header.xs}
                sm={this.props.layout.header.sm}
                md={this.props.layout.header.md}
                lg={this.props.layout.header.lg}
                className="strong text-start pt-2"
                data-tip
                data-for="headerTip"
              >
                <Image className="pe-2" src={this.props.headerLogo}></Image>
                {this.props.header}
                <ReactTooltip id="headerTip" place="bottom" effect="solid">
                  {this.props.header}
                </ReactTooltip>
              </Col>
              <Col
                xs={this.props.layout.dropdown.xs}
                sm={this.props.layout.dropdown.sm}
                md={this.props.layout.dropdown.md}
                lg={this.props.layout.dropdown.lg}
              >
                <Form.Select
                  className="dropdown-select"
                  aria-label="Default select example"
                  onChange={this.props.dropdownHandler}
                >
                  {this.props.reportTableConfig.dropdownData?.map(
                    (item: any) => (
                      <option key={item.id}>{item.name}</option>
                    )
                  )}
                </Form.Select>
              </Col>
              <Col
                xs={this.props.layout.viewAll.xs}
                sm={this.props.layout.viewAll.sm}
                md={this.props.layout.viewAll.md}
                lg={this.props.layout.viewAll.lg}
                className="pt-2"
              >
                <FormLabel
                  onClick={() => window.open("", "_blank")}
                  style={{ cursor: "pointer" }}
                  className="mylink"
                >
                  {this.props.localeConfig.reportTable_viewAll}
                </FormLabel>
              </Col>
            </Row>

            <Col
              id="agGridRoportTable"
              className="ag-theme-alpine"
              xs={this.props.layout.table.xs}
              sm={this.props.layout.table.sm}
              md={this.props.layout.table.md}
              lg={this.props.layout.table.lg}
              style={{ height: `${this.props.height ?? this.tableHeight()}px` }}
            >
              <AgGridReact
                getRowClass={getRowClass}
                suppressRowHoverHighlight={true}
                onCellClicked={(CellClickedEvent) =>
                  this.rowClickHandler(CellClickedEvent)
                }
                rowData={this.state.reportTableConfig.rowData}
                columnDefs={this.state.reportTableConfig.columnDefs}
                rowSelection={this.state.reportTableConfig.rowSelection}
                enableBrowserTooltips={true}
              />
            </Col>
          </Row>
        )}
      </>
    );
  }
}

export default ReportTable;
export { ReportTable };
