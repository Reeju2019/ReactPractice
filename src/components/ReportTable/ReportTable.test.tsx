import { shallow } from "enzyme"
import ReactDOM from "react-dom"
import ReportTable from "./ReportTable"

const localeConfig = {
  uploadAsset_header_1: "Upload your",
  uploadAsset_header_2: "Asset",
  uploadAsset_subHeader_1: "Supported File Formats:",
  uploadAsset_subHeader_2: "img, jpeg, png and etc",
  uploadAsset_dropBoxHeader_1: "Drop your files here or",
  uploadAsset_dropBoxHeader_2: "Browse",
  uploadAsset_dropBoxSubHeader_1: "Max size of Video: 25 MB",
  uploadAsset_upload_msg_success: "uploading Complete!",
  uploadAsset_upload_msg_failure:
    "something went wrong, please check file type/size and upload again!",
}

const data = {
  keys: ["Course Name", "Enrollments", "Efficacy"],
  rowSelection: "multiple",
  columnDefs: [
    {
      headerName: "Course Name",
      field: "Course Name",
      sortable: true,
      filter: true,
      checkboxSelection: false,
    },
    {
      headerName: "Enrollments",
      field: "Enrollments",
      sortable: true,
      filter: true,
      checkboxSelection: false,
    },
    {
      headerName: "Efficacy",
      field: "Efficacy",
      sortable: true,
      filter: true,
      checkboxSelection: false,
    },
  ],
  rowData: [
    { "Course Name": "Course Title 1", Enrollments: 53, Efficacy: 50 },
    { "Course Name": "Course Title 2", Enrollments: 24, Efficacy: 78 },
    { "Course Name": "Course Title 3", Enrollments: 21, Efficacy: 20 },
    { "Course Name": "Course Title 4", Enrollments: 18, Efficacy: 56 },
    { "Course Name": "Course Title 5", Enrollments: 25, Efficacy: 24 },
  ],
}

describe("App Testing", () => {
  it("render without crashing", () => {
    const div = document.createElement("div")
    ReactDOM.render(
      <ReportTable
        isLink={false}
        localeConfig={localeConfig}
        reportTableConfig={data}
        viewDetails={() => {}}
        header={"header"}
        headerLogo={"/assets/logo/example.png"}
        layout={{
          dropdown: {
            xs: 3,
            sm: 3,
            md: 3,
            lg: 2,
          },
          header: {
            xs: 8,
            sm: 8,
            md: 6,
            lg: 3,
          },
          viewAll: {
            xs: 4,
            sm: 4,
            md: 4,
            lg: 4,
          },
          table: {
            xs: 12,
            sm: 12,
            md: 12,
            lg: 12,
          },
        }}
        height={500}
        dropdownHandler={() => {}}
      />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })
  it("should render initial layout", () => {
    const component = shallow(
      <ReportTable
        isLink={false}
        localeConfig={localeConfig}
        reportTableConfig={data}
        viewDetails={() => {}}
        header={"header"}
        headerLogo={"/assets/logo/example.png"}
        layout={{
          dropdown: {
            xs: 3,
            sm: 3,
            md: 3,
            lg: 2,
          },
          header: {
            xs: 8,
            sm: 8,
            md: 6,
            lg: 3,
          },
          viewAll: {
            xs: 4,
            sm: 4,
            md: 4,
            lg: 4,
          },
          table: {
            xs: 12,
            sm: 12,
            md: 12,
            lg: 12,
          },
        }}
        height={500}
        dropdownHandler={() => {}}
      />
    )
    expect(component.getElements()).toMatchSnapshot()
  })
})
