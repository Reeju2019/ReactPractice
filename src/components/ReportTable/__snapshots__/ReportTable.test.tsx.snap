// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Testing should render initial layout 1`] = `
Array [
  <React.Fragment>
    <Row
      className="rchart-container-border m-5 m-5"
    >
      <Row
        className="rchart-border-bottom py-2 ps-4"
      >
        <Col
          className="strong text-start pt-2"
          data-for="headerTip"
          data-tip={true}
          lg={3}
          md={6}
          sm={8}
          xs={8}
        >
          <Image
            className="pe-2"
            fluid={false}
            rounded={false}
            roundedCircle={false}
            src="/assets/logo/example.png"
            thumbnail={false}
          />
          header
          <ReactTooltip
            clickable={false}
            effect="solid"
            id="headerTip"
            insecure={true}
            place="bottom"
            resizeHide={true}
            wrapper="div"
          >
            header
          </ReactTooltip>
        </Col>
        <Col
          lg={2}
          md={3}
          sm={3}
          xs={3}
        >
          <FormSelect
            aria-label="Default select example"
            className="dropdown-select"
            onChange={[Function]}
          />
        </Col>
        <Col
          className="pt-2"
          lg={4}
          md={4}
          sm={4}
          xs={4}
        >
          <FormLabel
            className="mylink"
            column={false}
            onClick={[Function]}
            visuallyHidden={false}
          />
        </Col>
      </Row>
      <Col
        className="ag-theme-alpine"
        id="agGridRoportTable"
        lg={12}
        md={12}
        sm={12}
        style={
          Object {
            "height": "500px",
          }
        }
        xs={12}
      >
        <AgGridReact
          columnDefs={
            Array [
              Object {
                "checkboxSelection": false,
                "field": "Course Name",
                "filter": true,
                "headerName": "Course Name",
                "headerTooltip": "Course Name",
                "sortable": true,
              },
              Object {
                "checkboxSelection": false,
                "field": "Enrollments",
                "filter": true,
                "headerName": "Enrollments",
                "headerTooltip": "Enrollments",
                "sortable": true,
              },
              Object {
                "checkboxSelection": false,
                "field": "Efficacy",
                "filter": true,
                "headerName": "Efficacy",
                "headerTooltip": "Efficacy",
                "sortable": true,
              },
            ]
          }
          enableBrowserTooltips={true}
          getRowClass={[Function]}
          onCellClicked={[Function]}
          rowData={
            Array [
              Object {
                "Course Name": "Course Title 1",
                "Efficacy": 50,
                "Enrollments": 53,
              },
              Object {
                "Course Name": "Course Title 2",
                "Efficacy": 78,
                "Enrollments": 24,
              },
              Object {
                "Course Name": "Course Title 3",
                "Efficacy": 20,
                "Enrollments": 21,
              },
              Object {
                "Course Name": "Course Title 4",
                "Efficacy": 56,
                "Enrollments": 18,
              },
              Object {
                "Course Name": "Course Title 5",
                "Efficacy": 24,
                "Enrollments": 25,
              },
            ]
          }
          rowSelection="multiple"
          suppressRowHoverHighlight={true}
        />
      </Col>
    </Row>
  </React.Fragment>,
]
`;
